docs/adr/006_cups.md

```diff
@@ -0,0 +1,297 @@
+Function->>IoT Hub: Retrieve Concentrator device Twin
+Function->>Function: Parse *CredentialUrl string to understand credentials location
+alt *CredentialUrl is a KeyVault Secret
+    Function->>Azure KeyVault: Get KeyVault secret using Azure Function Managed Identity
+else *CredentialUrl is from Azure Blob Storage and no SAS is in Url
+    Function->>Azure Blob Storage: Get Blob using MSI
+else *CredentialUrl is plain Url
+    Function->>HTTPS Endpoint: Get Blob with HTTP GET Request
+end
+Function->>LoRaWanNetworkSrvModule: Return credential blob
+Note right of LoRaWanNetworkSrvModule: Response should be a json including base64 encoded credential
+
+```
+
+In addition to this, the Azure Function must be able to properly authenticate to target sink via Managed Identity (when using Azure Blob Storage or KeyVault).
+
+Considering that the default sink for the starter kit should be an Azure Storage Account, the template for the starter kit should be changed in such a way that a new Blob Container is created for uploading credential blob files.
+
+If using KeyVault secrets instead of Blob Storage, instructions on how to create a role assignment should be provided.  
+
+### LoRaWan Network Server changes
+
+- LoRaDeviceAPIServiceBase and its implementation (LoRaDeviceAPIService) should include a method for calling the 'FetchConcentratorCredentials' endpoint
+- IBasicsStationConfigurationService and its implementation (BasicsStationConfigurationService) should include a method for retrieving and parsing "cups" desired property in concentrator device twin
+- BasicsStationNetworkServerStartup should include a new endpoint for handling CUPS '/update-info' POST requests
+- BasicsStationNetworkServer should implement a basic client certificate validation when configuring kestrel
+- A new CupsProtocolMessageProcessor class is implemented for achieving purposes of above drawn sequence diagram.
+
+### LoRa Device Provisioning CLI changes
+
+The CLI should be changed in order to:
+
+- Specify if the device being created is making use of CUPS or not
+- In case a device being created is making use of CUPS, the tool should:
+  - accept as input a cupsCredentials file (being the concatenation of `cups.{trust,cert,key}`)
+  - accept as input a tcCredentials file (being the concatenation of `tc.{trust,cert,key}`)
+  - accept as input a clientThumbprint string (for client certificate validation)
+  - compute the CRC32 of above mentioned files
+  - create a device for the concentrator
+  - upload, via Azure Blob Storage C# SDK, the credential files to the Blob Storage
+  - update the Twin for the concentrator device accordingly
+- In case a device being created is not making use of CUPS, tool should provide a `--no-cups` option, allowing to just specify a `clientThumbprint` (for client certificate validation on LNS endpoint)
+- In case a device is already created and a credential rotation is required, the tool should:
```

Comment:
How do we plan to support rotation of LNS certs? Since the CUPS endpoint will be part of the same problem, we could have a chicken-and-egg problem there if the client wants to connect to the HTTPS CUPS endpoint, no?

Reply:
That's a good point, I was wondering of having the cups server directly on the function, but that would pose a series of other issues.

Reply:
Yeah I thought as much. I think "we won't do this" is also a fair answer, will resolve the comment.

Reply:
Both of you are right, rotation of root certificate should not happen frequently, therefore we can consider this an edge scenario.

That said, the idea would be the following:
- Another LoRaNetworkServer instance should be executed in parallel (with different PORT ideally) with the updated Server Certificate.
- Provisioning CLI tool should update all the concentrator devices with new CUPS URI and the new certificate bundles (which are now including the updated *.trust). The update process would be the same as it was a simple client certificate rotation process.
- As soon as a device is connecting to the "old" CUPS endpoint, a new uri and credentials bundle will be provided to the station.
- The Station will restart and try to use the "new" CUPS endpoint.
- Ideally after 24hrs, it's possible to remove the "old" LoRaNetworkServer instance

If you agree, I could include this scenario in the ADR or in the upcoming documentation.

@bastbu, @Mandur, any thoughts?